### PR TITLE
Remove blocking from RSH#deleteRetentionLease

### PR DESCRIPTION
**Pull request by DaveCTurner**
Similarly to #95309, we can acquire the primary operation permit without blocking a thread.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@original-brownbear this is ready to go IMO, I've had `./gradlew :server:test :server:internalclustertest` running in a loop overnight with no issues.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -999,6 +986,26 @@ void createRetentionLease(final long startingSeqNo, ActionListener<RetentionLeas
...
+            shard.getThreadPool().generic(),
+            outerListener.<Void>delegateFailure((l, ignored) -> leaseListener.addListener(l))
+        );
+
+        runUnderPrimaryPermit(permitListener -> ActionListener.completeWith(permitListener, () -> {
+            try {
+                shard.removePeerRecoveryRetentionLease(request.targetNode().getId(), delayedListener.map(ignored -> null));
+            } catch (RetentionLeaseNotFoundException e) {
+                // this counts as success
+                logger.debug("no peer-recovery retention lease for " + request.targetAllocationId());
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:1002]: NIT :) Not that it really matters and it was like that before but maybe remove string concat from logger invocation here?

**Review by original-brownbear — APPROVED** [review]
LGTM

side comment: ... the listener stuff is getting a little out of control isn't it? `deleteRetentionLease` is incredibly hard to read IMO but I admit that we don't really have any tooling that would make it look nicer at this point ...

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Ah yep ok, see 87ca673238a

**Comment by DaveCTurner** [comment]
> we don't really have any tooling that would make it look nicer

I'll have a think about this and maybe open a follow-up.

**Pull request merged**